File: Java/math/sum_of_divisors.java
/*
Given a Number. Find it's Sum of Divisors.
The Number can be big such as 10^6 to 10^9
So, If we do this in O(N) we will get Time Limit Exceeded [TLE]
Better Approach is run a loop till sqrt(Number)
Example square root of 10^9 is 31622 (approx), so iteration optimized a lot.
*/

import java.util.Scanner;
import java.lang.*;
import java.math.*;

public class SumofDivisors
{
   //this get_sum_of_divisors will count the sum of divisors of the number
   static long get_sum_of_divisors(long number)
   {
    /* Sum of Divisors may not fit in int range
       so better take long long int */
      long sum_of_divisors = 0, i = 0;
      int limit = (int) Math.sqrt(number);
      for(i = 1; i < limit; i++)
      {
         if(number % i == 0)
         {
            /* Here i is a divisor of that number
            ( Number / i ) is also a divisor
            */
            sum_of_divisors += i;
            sum_of_divisors += (number / i);
         }
      }
	

    if(number % limit == 0)
    {
        if(limit * limit == number)
        {
            //means perfect square number
            sum_of_divisors += limit;
        }
        else
        {
            sum_of_divisors += limit;
            sum_of_divisors += (number / limit);
        }
    }
    return sum_of_divisors;
	   
   }

    public static void main(String args[])  
    {
        Scanner scan = new Scanner(System.in);
	System.out.print("Enter the number : \n");
	long number = scan.nextLong();
        long sum_of_divisors = get_sum_of_divisors(number);
        
        System.out.print("Sum of Divisors of this Number is : ");
        System.out.print(sum_of_divisors);
        scan.close();
    }
}
/*

Standard Input and Output

Enter the number :
1000000000
Sum of Divisors of this Number is : 2497558338

Enter the number :
456723
Sum of Divisors of this Number is : 681408

Time Complexity : O( sqrt(N) )
Space Complexity : O( 1 )

*/

File: C-Plus-Plus/other/Sum_of_divisors.cpp
/*
Sum of all the possible divisors of a natural number

You will be given a natural number and your task is
to find all possible divisors of the given number and
return their sum as the output.
*/

#include <bits/stdc++.h>

using namespace std;

//function to sum up all the possible divisors of a natural number
int divisorSum(int number)
{
    int ans = 0;
    for(int pass = 2; pass <= sqrt(number); pass++)
    {
        //for i is a divisor then remainder must be 0
        if(number % pass == 0)
        {
            //if the divisors are same we count only one of 
            //them in sum
            if(pass == (number / pass))
            {
                ans = ans + pass;
            }
            //otherwise we count both for sum
            else
            {
                ans = ans + (pass + (number / pass));
            }
        }
    }
    return (ans + 1);
}

//driver code
int main()
{
    int number;
    cout << "Enter a natural number: ";
    cin >> number;
    cout << "Sum of all possible divisors of " << number << " is: " << divisorSum(number);
    return 0;
}

/*
EXAMPLE:-

Example 1:-
Input--
Enter a natural number: 40
Output--
Sum of all possible divisors of 40 is: 90

Example 2:-
Input--
Enter a natural number: 1
Output--
Sum of all possible divisors of 1 is: 1

TIME COMPLEXITY--> O(sqrt(N))
SPACE COMPLEXITY--> O(1)
*/
